src/clients/SpokePoolClient.ts

```diff
@@ -725,11 +725,17 @@ export class SpokePoolClient extends BaseAbstractClient {
       for (const event of speedUpEvents) {
         const speedUp: SpeedUp = { ...spreadEvent(event.args), originChainId: this.chainId };
         assign(this.speedUps, [speedUp.depositor, speedUp.depositId], [speedUp]);
-      }
 
-      // Traverse all deposit events and update them with associated speedups, If they exist.
-      for (const deposits of Object.entries(this.depositHashes)) {
-        this.depositHashes[deposits[0]] = this.appendMaxSpeedUpSignatureToDeposit(deposits[1]);
+        // Find deposit hash matching this speed up event and update the deposit data associated with the hash,
+        // if the hash+data exists.
+        const depositHash = this.getDepositHash(speedUp);
+
+        // We can assume all deposits in this lookback window are loaded in-memory already so if the depositHash
+        // is not mapped to a deposit, then we can throw away the speedup as it can't be applied to anything.
+        const depositDataAssociatedWithSpeedUp = this.depositHashes[depositHash];
+        if (isDefined(depositDataAssociatedWithSpeedUp)) {
+          this.depositHashes[depositHash] = this.appendMaxSpeedUpSignatureToDeposit(depositDataAssociatedWithSpeedUp);
+        }
```

Comment:
Given that we are no longer being mindful of a looping mechanism, this looks good. And, since we build forward in time, we shouldn't run into the case where a deposit is loaded at a later call to `update()` and after its `speedUp` events are retrieved.